src/LSST/control/Context.cpp

```diff
@@ -38,12 +39,15 @@ Context::Ptr Context::_thisPtr;
 std::mutex Context::_thisMtx;
 
 void Context::setup() {
-    lock_guard<mutex> lock(_thisMtx);
-    if (_thisPtr) {
-        LERROR("Config already setup");
-        return;
+    {
+        lock_guard<mutex> lock(_thisMtx);
+        if (_thisPtr) {
+            LERROR("Config already setup");
+            return;
+        }
+        _thisPtr = Ptr(new Context());
     }
-    _thisPtr = Ptr(new Context());
+    _thisPtr->model.getPowerSystem()->setContext(_thisPtr);

```

Comment:
It looks like there is some cyclic indirect dependency:
```
Context -> Model -> PowerSystem -> Context
```
This by itself may not be a problem. Though, in general, it's hard to comprehend and debug the code that has this. A problem would be if there is some instance that needs to be "tuned" with a pointer to `Context` before it could do anything. before Perhaps, there is a better way to design classes to avoid this? For example, one could construct things in the following sequence:
* `Context::setup` creates `Context` (that also creates `Model` if that's needed)
* `Context::setup` creates `PowerSystem` and passes a pointer to `Context` into c-tor of `PowerSystem`
The idea here is to avoid "micromanaging" objects by fine-tuning their states via scattered "setter" methods and pass as much information as needed to the c-tors of the classes instead.

Reply:
I don't really see any way to avoid the cyclic dependency. It's absolutely critical that the power system can inform the Model when the state changes and equally critical that the Model can use the PowerSystem to turn the power on and off. 

I think making the PowerSystem a singleton will probably make it look nicer (and avoid accidentally defining more than one PowerSystem), but it will probably make test_PowerSystem.cpp significantly uglier. It's a big enough change that it should be done in a new ticket.